implemented pagination in blog website

michael-020 opened the pull request:
Fixes issue : [keploy/issues/2546](https://github.com/keploy/keploy/issues/2546)

I have deployed my forked repository on vercel to test if the issue is really solved as the latest blog posts were got getting fetch in production.
My deployed url: [https://blog-website-psi-green.vercel.app/blog/community](https://blog-website-psi-green.vercel.app/blog/community)
#### Please Note: ensure you are on '/blog/community' route/endpoint while reviewing as there is no page '/' route 

Here's the screenshot of how the deployed application looks like:

<img width="1427" alt="Image" src="https://github.com/user-attachments/assets/8ae38a11-7b13-434a-81f6-e99283d5fcf3" />

manasmanohar commented:
Thanks for the raising the PR! @michael-020 

Could you please incorporate these changes:
- Ensure to include before and after video.
- Ensure to utilize pagination feature in the api, instead of fetching all 50 posts at once.
- load more button( [reference](https://engineering.fb.com/))
- Ensure the blog card (fixed size for all cards) and its content's (title, image , author, etc) spacing and alignment are consistent with respect to each other.
![image](https://github.com/user-attachments/assets/b7750739-ffff-423a-84ad-27464c5df098)
![image](https://github.com/user-attachments/assets/b85cad99-9b38-4462-9bc3-82ea67bdd79e)

- Ensure new blog cards appear smoothly without an abrupt flash.

michael-020 commented:
Hi @manasmanohar, I have implemented the first 2 points you mention, please give me some more time I am working on ensuring consistent alignments of elements in cards.
Currently, after clicking on "Load More Posts" I am rendering the next 12 posts, but if you would like me to change the number of posts being rendered then please let me know

https://github.com/user-attachments/assets/8250e445-4eef-4f5c-ac3e-fc5c747ff7a9

michael-020 commented:
Hi @manasmanohar @Sonichigo I have made the necessary changes, and I didn't find any of the elements in the cards being mis-aligned, I have added setTimeouts here to add a little bit of delay to create loading effect but I can remove it if it feels wrong. Please review the changes when you get a chance. Thanks!

manasmanohar commented:
Your pagination approach is closer to what we need, but two critical issues need addressing:
- Our API functions (getAllPostsForCommunity, getAllPostsForTechnology) don't currently support the pagination parameters you're trying to use - they need modification.
- Ensure to update the relevant pages with pagination

Currently, we are loading the entire list of posts on the frontend. Ideally, we should fetch the first 21 posts and then load the next set by clicking the “Load More” button.

michael-020 commented:
> Your pagination approach is closer to what we need, but two critical issues need addressing:
> 
> * Our API functions (getAllPostsForCommunity, getAllPostsForTechnology) don't currently support the pagination parameters you're trying to use - they need modification.
> * Ensure to update the relevant pages with pagination
> 
> Currently, we are loading the entire list of posts on the frontend. Ideally, we should fetch the first 21 posts and then load the next set by clicking the “Load More” button.

I am thinking of modifying the API functions to fetch 21 posts at once(currently it fetches entire list of posts), so that everytime a user clicks on the "Load More Posts" button a set of more 21 posts are fetched and then render only 12 and keep other 7 in reserve to set a consistent pattern of rendering posts in batches of 12, in this way I will be able to implement what is asked.
@manasmanohar Please let me know what you think of this approach, should I make any more modifications?

manasmanohar commented:
@michael-020 sounds good!

michael-020 commented:
Hi @manasmanohar , Please take a look at the video now I have implemented pagination by modifying the API function, but I have hardcoded the Wordpress url in the lib/api.ts file `const API_URL = process.env.WORDPRESS_API_URL || "https://wp.keploy.io/graphql";`
without doing this the re-fetch for pagination isn't working, I have put the Wordpress url in the .env file as well but I guess it is not getting read for some reason. Should I push the changes or remove the hardcoded wordpress url and then push it?

Video:

https://github.com/user-attachments/assets/1d42a63b-421c-446b-bb40-823bd6a03806

manasmanohar commented:
Hi @michael-020, did you add the .env file in the root of the project? Hard coding is not recommended. Will have to see the changes to debug.

michael-020 commented:
Hi @manasmanohar  I did add the .env in the root of the project just as mention in the readme.md file, but it still doesn't work for me but when I change the variable name to `NEXT_PUBLIC_WORDPRESS_API_URL=https://wp.keploy.io/graphql` it works, but for now I have removed the hardcoded wordpress url in my code and pushed the changes please check it out.

michael-020 commented:
Fixes: [keploy/issues/2546](https://github.com/keploy/keploy/issues/2546)

## Description
- Implemented pagination in blog-website.
- Fetch 21 posts initially and display them in batches of 12.


### Changes
- Modified API function to fetch 21 posts initially, then load more posts as a user click on the "Load More" button.
- The posts are displayed in batches of 12.
- Modified the more-stories component which displays all the blog posts to ensure pagination.


## Type of Change
- [x] Chore (maintenance, refactoring, tooling updates)
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (change that adds functionality)
- [ ] Breaking Change (may require updates in existing code)
- [x] UI improvement (visual or design changes)
- [x] Performance improvement (optimization or efficiency enhancements)
- [ ] Documentation update (changes to README, guides, etc.)
- [ ] CI (updates to continuous integration workflows)
- [ ] Revert (undo a previous commit or merge)


## Testing
- I logged the posts to verify that 21 posts are getting fetched at each API call.
- I looked at the network tab in developer tools of my browser to verify if API calls are working correctly.
- Ensured contents inside the card component are aligned properly.


## Demo

https://github.com/user-attachments/assets/1d42a63b-421c-446b-bb40-823bd6a03806


## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added corresponding tests
- [x] I have run the build command to ensure there are no build errors
- [x] My changes have been tested across relevant browsers/devices
- [x] For UI changes, I've included visual evidence of my changes

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -14,15 +15,64 @@ export default function MoreStories({
+      setLoading(true);
+      setError(null);
+      try {
+        const result = await getAllPostsForCommunity(false);
```
how are we requesting the next set of posts here? pass the required pagination parameters
how are we tracking pagination state, endCursor?
same set of posts are being fetched again

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -14,15 +15,64 @@ export default function MoreStories({
+    node.excerpt.toLowerCase().includes(searchTerm.toLowerCase())
+  );
+
+  console.log("api url: ", API_URL)
```
remove debug logs

manasmanohar commented on .github/PULL_REQUEST_TEMPLATE.md:
this file is already present in main. please ensure your fork is updated with latest changes from main.

manasmanohar commented:
Hey @michael-020, 
Please incorporate these changes. Let me know incase you have any query. 
1. In Next.js, environment variables that need to be exposed to the client must have the NEXT_PUBLIC_ prefix. Previously, with our SSG-based setup, regular environment variables were sufficient. However, since we are now fetching new posts on the client side, we need to expose the API URL to the client. By adding the NEXT_PUBLIC_ prefix, we can ensure that the variable is accessible both during local testing and in the production build. You can add this env our build file. 

2. Please recheck the pagination implementation, we need to track the endCursor from prev request, and pass the  pagination parameters to the api to get the next set of posts. Currently upon clicking load more, getting the same set of posts again. 

https://github.com/user-attachments/assets/0108148c-38a4-4d21-a8b1-70f78cddbf73

michael-020 commented:
Hi @manasmanohar, I have done the following things:

- Remove unused pagination variables from getAllPostsForCommunity function
- Correctly utilize the 'after' parameter for cursor-based pagination
- Ensure pageInfo (hasNextPage, endCursor) is returned from API
- Fix duplicate post loading in MoreStories component

The changes ensure that clicking "Load More" fetches new posts instead 
of repeating the same set of posts. 
Please review the changes when you get a chance. Thanks!

manasmanohar commented:
Hey @michael-020, you are heading in the right direction! Just a few adjustments needed:

- Instead of reusing the SSG functions (getAllPostsForCommunity and getAllPostsForTechnology) for pagination, consider adding a dedicated fetchMorePosts function to handle the client-side "Load More" functionality. This separates concerns between SSG and client-side pagination, making the code cleaner and more maintainable.
- I noticed that while your modified API functions now correctly return both edges and pageInfo, the page components (community/index.tsx and technology/index.tsx) are destructuring only the edges in their function parameters: ({ allPosts: { edges }, preview }). This causes the crucial pageInfo to be lost before reaching the MoreStories component.
- The MoreStories component currently initializes endCursor to null, which means when the "Load More" button is clicked, it will fetch the first batch of posts again rather than continuing from where the SSG left off. Make sure to pass the initial pageInfo down to MoreStories.
- Your idea about implementing  buffer strategy (fetching 21 posts, showing 12) is excellent for UX! However, the current implementation doesn't fully leverage this approach. Consider showing the buffer posts immediately when "Load More" is clicked, while fetching the next batch in the background.

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -14,15 +15,66 @@ export default function MoreStories({
+          setHasMore(result.pageInfo.hasNextPage);
+          
+          // Increase visible count
+          setVisibleCount(prev => prev + 12);
```
how are you we utilizing the buffer here?  currently when the user click on load more, it immediately fetches more post from api, leaving the user in loading state every time. the load more post could show post from buffer and then fetch more in background.

instead of the 21 posts, we could 
- fetch 22 posts initially (1 hero + 12 grid + 9 buffer)
- this gives you 3 complete rows (9 posts) in buffer
- when load more is clicked, add exactly 9 posts (3 complete rows) and fetch the next set of post in background.

and ensure we don't show load more posts button when there is no more post to fetch.

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -45,26 +97,50 @@ export default function MoreStories({
       )}
 
+      {error && (
+        <div className="text-red-500 text-center mb-8 p-4 bg-red-50 rounded-lg">
```
do add a screenshot of how this error would appear.

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -14,15 +15,66 @@ export default function MoreStories({
+      setLoading(true);
+      setError(null);
+      try {
+        const result = await getAllPostsForCommunity(false, endCursor);
```
more-stories component is a reusable components in our codebase can be used in diff pages. instead of hardcoding the function call or category, we could either utilize the existing prop or pass a category prop to utilize it for diff pages.

manasmanohar commented on components/more-stories.tsx:
```diff
@@ -14,15 +15,66 @@ export default function MoreStories({
+  const [loading, setLoading] = useState(false);
+  const [hasMore, setHasMore] = useState(true);
+  const [error, setError] = useState(null);
+  const [endCursor, setEndCursor] = useState(null);
```
without initial pageInfo, the first load more will fetch the first page again, causing duplicate posts. ensure to pass the initial pageInfo to it.

manasmanohar commented on lib/api.ts:
```diff
@@ -1,7 +1,7 @@
 export const dynamic = 'force-dynamic';
 
-const API_URL = process.env.WORDPRESS_API_URL;
+export const API_URL = process.env.WORDPRESS_API_URL;
```
why do we export the api url here?  to ensure client side calls are able to access the wordpress api url add and also add it in build step
```const API_URL = process.env.WORDPRESS_API_URL || process.env.NEXT_PUBLIC_WORDPRESS_API_URL```

manasmanohar commented on .github/PULL_REQUEST_TEMPLATE.md:
this file serves as our pull request template for github. you can remove this and use the template directly through the pull request page in the web ui. for future changes, copy and paste the template into a comment as needed

manasmanohar commented on pages/community/index.tsx:
```diff
@@ -55,7 +55,10 @@ export const getStaticProps: GetStaticProps = async ({ preview = false }) => {
-    props: { allPosts, preview },
+    props: { 
+      allPosts, 
+      preview
```
unless formatting brings significant improvement, ignore

manasmanohar commented on lib/api.ts:
```diff
@@ -309,16 +309,19 @@ export async function getAllPostsForTechnology(preview) {
   return { edges: allEdges };
 }
 
-export async function getAllPostsForCommunity(preview) {
```
itt would be better to utilize a separate fetchMorePosts function that can be used for both the community and technology pages to implement client-side pagination. this approach helps separate concerns between ssg functionality and client-side pagination. additionally, we can remove the while loop and send fewer posts instead of entire posts.

manasmanohar commented on lib/api.ts:
```diff
@@ -355,26 +358,25 @@ export async function getAllPostsForCommunity(preview) {
+  
+  return {
+    edges: data?.posts?.edges || [],
+    pageInfo: data?.posts?.pageInfo || { hasNextPage: false, endCursor: null }
```
ensure we update the pages to utilize the pageInfo, so that we can pass this info to the moreStories component inorder to  continue fetching posts from where ssg stopped.

michael-020 commented:
Fixes: [keploy/issues/2546](https://github.com/keploy/keploy/issues/2546)

## Description
- Implemented client-side pagination in the blog website
- Separated SSG and client-side data fetching concerns
- Added proper TypeScript types for pagination data structures

### Changes
1. **API Layer**:
   - Added dedicated `fetchMorePosts` function in `api.ts` for client-side pagination
   - Modified existing API functions to return both `edges` and `pageInfo`

2. **Components**:
   - Updated `MoreStories` component to handle pagination state
   - Added buffer-based loading for smooth UX
   - Implemented proper initialization of pagination state from SSG data

3. **Pages**:
   - Updated `getStaticProps` in all pages to include `pageInfo`
   - Fixed prop destructuring to properly pass `pageInfo` to child components

## Type of Change
- [x] Chore (maintenance, refactoring, tooling updates)
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (change that adds functionality)
- [ ] Breaking Change (may require updates in existing code)
- [x] UI improvement (visual or design changes)
- [x] Performance improvement (optimization or efficiency enhancements)
- [ ] Documentation update (changes to README, guides, etc.)
- [ ] CI (updates to continuous integration workflows)
- [ ] Revert (undo a previous commit or merge)

## Testing
- Added comprehensive logging throughout the pagination flow
- Verified proper initialization of pagination state from SSG data
- Tested buffer-based loading mechanism
- Ensured proper type checking across components
- Verified smooth loading experience with buffer system

## Demo

https://github.com/user-attachments/assets/66978fed-6e54-48f1-b4e5-ad28c665a342

Logs for initial pageInfo and other details: 

<img width="891" alt="Screenshot 2025-03-25 at 5 40 07 PM" src="https://github.com/user-attachments/assets/f72048c7-8ee9-4355-8731-01351954d74f" />

This is how the error message would look like: 

<img width="1401" alt="Screenshot 2025-03-25 at 5 26 06 PM" src="https://github.com/user-attachments/assets/c3ee9486-9c6b-44f0-bbb7-bbb73c03b4c0" />

Made sure the "Load More posts" button is not visible if there are no more posts left:
<img width="1440" alt="Screenshot 2025-03-25 at 4 47 01 PM" src="https://github.com/user-attachments/assets/71678701-a7ea-4d35-9510-4b523bc6c50f" />

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added corresponding tests
- [x] I have run the build command to ensure there are no build errors
- [x] My changes have been tested across relevant browsers/devices
- [x] For UI changes, I've included visual evidence of my changes

michael-020 commented:
Hi @manasmanohar, I have pushed the new changes, hope this time all the requirements are met. Please review the changes when you get a chance. Thanks!

manasmanohar commented on pages/community/[slug].tsx:
```diff
@@ -213,24 +213,27 @@ export const getStaticProps: GetStaticProps = async ({
     },
-    revalidate: 10,
+    // CHANGE: Reduced revalidation time
+    revalidate: 60, // Revalidate every minute
```
remove the revalidation update.

manasmanohar commented:
Hi @michael-020, this is great! Few minor changes to resolve and we are good to go :
- Ensure the technology post and community page posts behaviour are the same. Currently the technology page is only showing 13 posts without a load more button for viewing remaining posts. Load more button should be shown given there are more posts. 

https://github.com/user-attachments/assets/b0b7c491-d2ca-4c4c-b856-67d4e714349e

- Fix the logic of error message element, display the error message element at the bottom while keeping the existing posts visible instead of hiding the already loaded posts.
- Remove the .github folder as its already present in main. you pr should include only the files to update.

michael-020 commented:
Fixes: [keploy/issues/2546](https://github.com/keploy/keploy/issues/2546)

## Description
- Implemented proper pagination in technology page.

### Changes
- Made changes so that the technology page mimics the behaviour of pagination as done my me previously in the community page.
- Fixed placement of error message element.

## Type of Change
- [x] Chore (maintenance, refactoring, tooling updates)
- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (change that adds functionality)
- [ ] Breaking Change (may require updates in existing code)
- [x] UI improvement (visual or design changes)
- [x] Performance improvement (optimization or efficiency enhancements)
- [ ] Documentation update (changes to README, guides, etc.)
- [ ] CI (updates to continuous integration workflows)
- [ ] Revert (undo a previous commit or merge)

## Testing
- Added comprehensive logging throughout the pagination flow
- Verified proper initialization of pagination state from SSG data
- Tested buffer-based loading mechanism
- Ensured proper type checking across components
- Verified smooth loading experience with buffer system
- Tested search functionality with pagination
- Verified error handling and UI feedback

## Demo

https://github.com/user-attachments/assets/ecdbc5be-6f6a-463d-9e1b-237815d5e642

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added corresponding tests
- [x] I have run the build command to ensure there are no build errors
- [x] My changes have been tested across relevant browsers/devices
- [x] For UI changes, I've included visual evidence of my changes

michael-020 commented:
Hi @manasmanohar, I have made changes in the technology page to work the same way as the community page. Please review the changes when you get a chance. Thanks!

manasmanohar commented:
There is scope for optimizing the functionality  and ux further. lets get this merged since we have few other dependancies on this pagination feature.

Hermione2408 commented:
LGTM!

michael-020 commented:
> LGTM!

Hi @Hermione2408,

I’m seeing the following error on https://keploy.io/blog/community. I encountered the same issue when the .env file had `WORDPRESS_API_URL=https://wp.keploy.io/graphql` instead of `NEXT_PUBLIC_WORDPRESS_API_URL=https://wp.keploy.io/graphql`.

Please ensure the .env file uses `NEXT_PUBLIC_WORDPRESS_API_URL=https://wp.keploy.io/graphql`, as Next.js recognizes environment variables with the `NEXT_PUBLIC_` prefix.
Thanks!

<img width="558" alt="Screenshot 2025-04-07 at 12 07 43 PM" src="https://github.com/user-attachments/assets/75d8270e-7979-4b99-a3f2-dbae01ab33b1" />
